test/unit/scan/test_scan_policies.py

```diff
@@ -7,14 +7,207 @@
 
 from apiver_deps import ScanPoliciesManager
 from apiver_deps_exception import InvalidArgument
 
 
+class TestFolderTraversal:
+    def test_simple_folder(self, tmp_path):
+        d = tmp_path / "dir"
+        d.mkdir()
+        (d / "file1.txt").write_text("content1")
+        (d / "file2.txt").write_text("content2")
+        (d / "file3.txt").write_text("content3")
+
+        folder = LocalFolder(str(d))
+        local_paths = folder.all_files(reporter=MagicMock())
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert len(absolute_paths) == 3
+        assert absolute_paths == [
+            fix_windows_path_limit(str(d / "file1.txt")),
+            fix_windows_path_limit(str(d / "file2.txt")),
+            fix_windows_path_limit(str(d / "file3.txt"))
+        ]
+
+    def test_folder_with_subfolders(self, tmp_path):
+        d1 = tmp_path / "dir1"
+        d1.mkdir()
+        (d1 / "file1.txt").write_text("content1")
+        (d1 / "file2.txt").write_text("content2")
+
+        d2 = tmp_path / "dir2"
+        d2.mkdir()
+        (d2 / "file3.txt").write_text("content3")
+        (d2 / "file4.txt").write_text("content4")
+
+        folder = LocalFolder(str(tmp_path))
+        local_paths = folder.all_files(reporter=MagicMock())
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert len(absolute_paths) == 4
+        assert absolute_paths == [
+            fix_windows_path_limit(str(d1 / "file1.txt")),
+            fix_windows_path_limit(str(d1 / "file2.txt")),
+            fix_windows_path_limit(str(d2 / "file3.txt")),
+            fix_windows_path_limit(str(d2 / "file4.txt"))
+        ]
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    def test_folder_with_symlink_to_file(self, tmp_path):
+        d = tmp_path / "dir"
+        d.mkdir()
+
+        file = d / "file.txt"
+        file.write_text("content")
+
+        symlink_file = tmp_path / "symlink_file.txt"
+        symlink_file.symlink_to(file)
+
+        folder = LocalFolder(str(tmp_path))
+        local_paths = folder.all_files(reporter=MagicMock())
+
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert len(absolute_paths) == 2
+        assert absolute_paths == [
+            fix_windows_path_limit(str(file)),
+            fix_windows_path_limit(str(symlink_file))
+        ]
+
+    #FIXME the following two tests could be combined to avoid code duplication
+    # but I decide to keep them separate for now to make it easier to debug
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    @pytest.mark.timeout(5)  # Set a 5-second timeout for this test
+    def test_folder_with_circular_symlink(self, tmp_path):
```

Comment:
I think we should recurse down 10 levels (or whatever the limit is) and then realize that we are in an infinite loop, raise an exception, catch it upstairs in the stack and see if the place where we are is an infinite loop already or have we went over it already and we are in a healthy directory again. If we are in a loop, raise again. Eventually we'll come all the way to the top where the _walk function fell into the loop and we'll realize that was a mistake - then we'll emit a warning (maybe tell the reporter what happened?) and skip.

If we wouldn't do that, we'd (as shown in this test) upload the same file (or files) ~10 times, wasting capacity. I do not think this is what the users of this class would want.